### PR TITLE
PEP 635: Add a rebuttal of PEP 642

### PR DESCRIPTION
**Pull request by brandtbucher**
CC @Tobias-Kohn

### CODE REVIEW

File: pep-0635.rst [hunk 1]
```diff
@@ -893,6 +893,34 @@ patterns impossible.
           case (HttpStatus.NOT_FOUND):
               raise ResourceNotFound()
 
+:pep:`642` (a competing document) proposes using ``?`` as a prefix for
```
Comment (gvanrossum) [inline pep-0635.rst:896]:
```suggestion
:pep:`642` proposes using ``?`` as a prefix for
```

File: pep-0635.rst [hunk 2]
```diff
@@ -893,6 +893,34 @@ patterns impossible.
...
+*constraint patterns*: arbitrary expressions that replace the value and literal
+patterns defined here.  The PEP is motivated by a desire to "treat match
+patterns as a variation on assignment targets".
+
+We believe that attempting to unify the grammars of assignments and patterns is
+attractive, but misguided.  Evidence of this is PEP 642's need to introduce new
+syntax (which *neither* requires) as a best attempt to unify them.  In contrast,
```
Comment (gvanrossum) [inline pep-0635.rst:903]:
```suggestion
syntax (which *neither* assignments nor our proposal requires) as a best attempt to unify them.  In contrast,
```

File: pep-0635.rst [hunk 3]
```diff
@@ -893,6 +893,34 @@ patterns impossible.
           case (HttpStatus.NOT_FOUND):
               raise ResourceNotFound()
 
+:pep:`642` (a competing document) proposes using ``?`` as a prefix for
+*constraint patterns*: arbitrary expressions that replace the value and literal
+patterns defined here.  The PEP is motivated by a desire to "treat match
+patterns as a variation on assignment targets".
+
+We believe that attempting to unify the grammars of assignments and patterns is
```
Comment (gvanrossum) [inline pep-0635.rst:901]:
```suggestion
We believe that attempting to unify the grammars of assignment targets and patterns is
```

File: pep-0635.rst [hunk 4]
```diff
@@ -893,6 +893,34 @@ patterns impossible.
...
+
+There are deeper issues as well.  Even if the grammars for both assignments and
+patterns are made "consistent" with one another, strings, byte-strings,
+mappings, sequences, and iterators will all behave differently in both contexts.
+This leaves us in a worse situation: one where the *grammar* is consistent, but
+the *behavior* differs in meaningful ways.
+
+The proposed syntax (prefixing arbitrary expressions with ``?``) is also quite
+ugly and distracting, harming readability.  (Our issues with using ``?`` as a
+token here have already been enumerated above.)
```
Comment (gvanrossum) [inline pep-0635.rst:922]:
```suggestion
wildcard symbol here have already been enumerated above.)
```

**Review by gvanrossum — COMMENTED** [review]
A few suggestions.

I find it telling that the counterproposals focus on `?`, which quite arbitrary (even more for PEP 642 than for PEP 640).

Possibly Nick just had to get this out of his system.

**Comment by Tobias-Kohn** [comment]
@brandtbucher and @gvanrossum: thank you very much for this.  I find your text really well written and to the point.

**Comment by gvanrossum** [comment]
All thanks are due to Brandt. I have added a mention of this in our issue for the SC. (https://github.com/python/steering-council/issues/39)

**Pull request merged**